## Data Mining 
### for a Global Data analysis of Human Environment in 162 countries in 2017
#### In this analysis should be examined the correlations between Freedom, Happiness, Religion, Safety, Life-expectancy, Gender-Equality, GDP_per_capita, Population, Peace, Freedom of speech, Corruption and Human development.
##### The Global Multidimensional Poverty Index (MPI) just includes 105 countries (77% of global population), so I decided not to use it.¶

##### To collect the features I choosed 9 datasets, all found by kaggle.com. The source of the Original-Data is given in every description of the single dataset in this notebook 

In [1]:
import pandas as pd

#### "The Human Freedom Index"
The Human Freedom Index is constructed by the Cato Institute, Fraser Institute, and the Friedrich Naumann Foundation for Freedom in 2018 and presents the state of human freedom in the world based on a broad measure of personal, civil, and economic freedom. On a scale of 0 to 10, where 10 represents more freedom, the average human freedom rating for 162 countries in 2017 was 6.89. 
Most important areas and their corresponding column names of this dataset mentioned as follows:

- Rule of Law (#pf_rol)
- Security and Safety (#pf_ss)
- Movement (#pf_movement)
- Religion (#pf_religion)
- Association, Assembly and Civil Society (#pf_association)
- Expression and Information (#pf_expression)
- Identity and Relationships (#pf_identity)
- Size of Government (#ef_government)
- Legal System and Property Rights (#ef_legal)
- Access to Sound Money (#ef_money)
- Freedom to Trade Internationally (#ef_trade)
- Regulation of Credit, Labor and Business (#ef_regulation)



##### Show all columns

In [2]:
pd.set_option("display.max_columns", None)

##### Import dataset 2008 - 2016

In [3]:
df_2016 = pd.read_csv("files/hfi_cc_2018.csv")
df_2016.head(3)

year ISO_code countries                      region  pf_rol_procedural  \
0  2016      ALB   Albania              Eastern Europe           6.661503   
1  2016      DZA   Algeria  Middle East & North Africa                NaN   
2  2016      AGO    Angola          Sub-Saharan Africa                NaN   

   pf_rol_civil  pf_rol_criminal    pf_rol  pf_ss_homicide  \
0      4.547244         4.666508  5.291752        8.920429   
1           NaN              NaN  3.819566        9.456254   
2           NaN              NaN  3.451814        8.060260   

   pf_ss_disappearances_disap  pf_ss_disappearances_violent  \
0                        10.0                      10.00000   
1                        10.0                       9.29403   
2                         5.0                      10.00000   

   pf_ss_disappearances_organized  pf_ss_disappearances_fatalities  \
0                            10.0                        10.000000   
1                             5.0                         9.926119   
2                             7.5                        10.000000   

   pf_ss_disappearances_injuries  pf_ss_disappearances  pf_ss_women_fgm  \
0                      10.000000              10.00000             10.0   
1                       9.990149               8.84206             10.0   
2                      10.000000               8.50000             10.0   

   pf_ss_women_missing  pf_ss_women_inheritance_widows  \
0                  7.5                             5.0   
1                  7.5                             0.0   
2                 10.0                             5.0   

   pf_ss_women_inheritance_daughters  pf_ss_women_inheritance  pf_ss_women  \
0                                5.0                      5.0     7.500000   
1                                0.0                      0.0     5.833333   
2                                5.0                      5.0     8.333333   

      pf_ss  pf_movement_domestic  pf_movement_foreign  pf_movement_women  \
0  8.806810                   5.0                 10.0                5.0   
1  8.043882                   5.0                  5.0                5.0   
2  8.297865                   0.0                  5.0               10.0   

   pf_movement  pf_religion_estop_establish  pf_religion_estop_operate  \
0     6.666667                          NaN                        NaN   
1     5.000000                          NaN                        NaN   
2     5.000000                          NaN                        NaN   

   pf_religion_estop  pf_religion_harassment  pf_religion_restrictions  \
0               10.0                9.566667                  8.011111   
1                5.0                6.873333                  2.961111   
2               10.0                8.904444                  7.455556   

   pf_religion  pf_association_association  pf_association_assembly  \
0     9.192593                        10.0                     10.0   
1     4.944815                         5.0                      5.0   
2     8.786667                         2.5                      2.5   

   pf_association_political_establish  pf_association_political_operate  \
0                                 NaN                               NaN   
1                                 NaN                               NaN   
2                                 NaN                               NaN   

   pf_association_political  pf_association_prof_establish  \
0                      10.0                            NaN   
1                       5.0                            NaN   
2                       2.5                            NaN   

   pf_association_prof_operate  pf_association_prof  \
0                          NaN                 10.0   
1                          NaN                  5.0   
2                          NaN                  5.0   

   pf_association_sport_establish  pf_association_sport_operate  \
0                             NaN     

In [4]:
df_2016.shape

(1458, 123)

##### Import dataset 2008 - 2017

In [5]:
df_2017 = pd.read_csv("files/hfi_cc_2019.csv")
df_2017.head(3)

year ISO_code countries                      region hf_score hf_rank  \
0  2017      ALB   Albania              Eastern Europe     7.84      38   
1  2017      DZA   Algeria  Middle East & North Africa     4.99     155   
2  2017      AGO    Angola          Sub-Saharan Africa      5.4     151   

  hf_quartile pf_rol_procedural pf_rol_civil pf_rol_criminal pf_rol  \
0           1               6.7          4.5             4.7    5.3   
1           4                 -            -               -    3.8   
2           4                 -            -               -    3.4   

  pf_ss_homicide pf_ss_disappearances_disap pf_ss_disappearances_violent  \
0            9.1                         10                           10   
1            9.5                         10                          9.5   
2            8.1                         10                          9.7   

  pf_ss_disappearances_organized pf_ss_disappearances_fatalities  \
0                             10                              10   
1                              5                             9.9   
2                            7.5                             9.9   

  pf_ss_disappearances_injuries pf_ss_disappearances pf_ss_women_fgm  \
0                            10                   10              10   
1                           9.9                  8.9              10   
2                           7.2                  8.9              10   

  pf_ss_women_inheritance_widows pf_ss_women_inheritance_daughters  \
0                              -                                 -   
1                              -                                 -   
2                              -                                 -   

  pf_ss_women_inheritance pf_ss_women pf_ss pf_movement_domestic  \
0                     7.5         8.8   9.3                   10   
1                       0           5   7.8                   10   
2                       5         7.5   8.1                    5   

  pf_movement_foreign pf_movement_women pf_movement  \
0                  10                10          10   
1                   5               2.5         5.8   
2                   5                10         6.7   

  pf_religion_estop_establish pf_religion_estop_operate pf_religion_estop  \
0                           -                         -                10   
1                           -                         -                 5   
2                           -                         -                10   

  pf_religion_harassment pf_religion_restrictions pf_religion  \
0                    9.6                        8         9.2   
1                    6.9                        3         4.9   
2                    8.9                      7.5         8.8   

  pf_association_association pf_association_assembly  \
0                         10                      10   
1                          5                       5   
2                        2.5                     2.5   

  pf_association_political_establish pf_association_political_operate  \
0                                  -                                -   
1                                  -                                -   
2                                  -                                -   

  pf_association_political pf_association_prof_establish  \
0                       10                             -   
1                        5                             -   
2                      2.5                             -   

  pf_association_prof_operate pf_association_prof  \
0                           -                  10   
1                           -                   5   
2                           -                   5   

  pf_association_sport_establish pf_association_sport_operate  \
0                              -                            -   
1                              -                            -   
2                              -      

In [6]:
df_2017.shape

(1620, 120)

#### Control, if Data 2008 - 2016 is same in both datasets

In [7]:
df_control1 = df_2016[["year", "countries", "hf_score", "hf_rank"]]
df_control2 = df_2017[["year", "countries", "hf_score", "hf_rank"]]
df_control1.head().round(2)

year  countries  hf_score  hf_rank
0  2016    Albania      7.57     48.0
1  2016    Algeria      5.14    155.0
2  2016     Angola      5.64    142.0
3  2016  Argentina      6.47    107.0
4  2016    Armenia      7.24     57.0

In [8]:
df_control2[162:167]

year  countries hf_score hf_rank
162  2016    Albania     7.67      45
163  2016    Algeria     5.22     153
164  2016     Angola     5.89     137
165  2016  Argentina     6.75      85
166  2016    Armenia     7.44      53

#### The Data is obviously touched up in the actual dataset. Argentina i. e. is ranked up for 22 places in 2016.
### For further exploration I use the 2008 - 2017 (df_2017) dataset

In [9]:
df_2017.head(3)

year ISO_code countries                      region hf_score hf_rank  \
0  2017      ALB   Albania              Eastern Europe     7.84      38   
1  2017      DZA   Algeria  Middle East & North Africa     4.99     155   
2  2017      AGO    Angola          Sub-Saharan Africa      5.4     151   

  hf_quartile pf_rol_procedural pf_rol_civil pf_rol_criminal pf_rol  \
0           1               6.7          4.5             4.7    5.3   
1           4                 -            -               -    3.8   
2           4                 -            -               -    3.4   

  pf_ss_homicide pf_ss_disappearances_disap pf_ss_disappearances_violent  \
0            9.1                         10                           10   
1            9.5                         10                          9.5   
2            8.1                         10                          9.7   

  pf_ss_disappearances_organized pf_ss_disappearances_fatalities  \
0                             10                              10   
1                              5                             9.9   
2                            7.5                             9.9   

  pf_ss_disappearances_injuries pf_ss_disappearances pf_ss_women_fgm  \
0                            10                   10              10   
1                           9.9                  8.9              10   
2                           7.2                  8.9              10   

  pf_ss_women_inheritance_widows pf_ss_women_inheritance_daughters  \
0                              -                                 -   
1                              -                                 -   
2                              -                                 -   

  pf_ss_women_inheritance pf_ss_women pf_ss pf_movement_domestic  \
0                     7.5         8.8   9.3                   10   
1                       0           5   7.8                   10   
2                       5         7.5   8.1                    5   

  pf_movement_foreign pf_movement_women pf_movement  \
0                  10                10          10   
1                   5               2.5         5.8   
2                   5                10         6.7   

  pf_religion_estop_establish pf_religion_estop_operate pf_religion_estop  \
0                           -                         -                10   
1                           -                         -                 5   
2                           -                         -                10   

  pf_religion_harassment pf_religion_restrictions pf_religion  \
0                    9.6                        8         9.2   
1                    6.9                        3         4.9   
2                    8.9                      7.5         8.8   

  pf_association_association pf_association_assembly  \
0                         10                      10   
1                          5                       5   
2                        2.5                     2.5   

  pf_association_political_establish pf_association_political_operate  \
0                                  -                                -   
1                                  -                                -   
2                                  -                                -   

  pf_association_political pf_association_prof_establish  \
0                       10                             -   
1                        5                             -   
2                      2.5                             -   

  pf_association_prof_operate pf_association_prof  \
0                           -                  10   
1                           -                   5   
2                           -                   5   

  pf_association_sport_establish pf_association_sport_operate  \
0                              -                            -   
1                              -                            -   
2                              -      

##### Looking for the Data-Types

In [10]:
df_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Columns: 120 entries, year to ef_rank
dtypes: int64(1), object(119)
memory usage: 1.5+ MB


#### Because the columns with numbers are all Strings (except "year" = int64), I create a function to change the Data-types to float64

In [11]:
a = list(df_2017.columns.values) # Create a columns-List
del a[1:4] # Delete the columns with no numbers ("ISO_code", "countries" and "region")
def changetype(i): # Function to change Datatype for every single column (116 columns has to be changed)
    for i in a:
        df_2017[i] = pd.to_numeric(df_2017[i], errors='coerce')
changetype(a)
df_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Columns: 120 entries, year to ef_rank
dtypes: float64(116), int64(1), object(3)
memory usage: 1.5+ MB


## Creating a Dataframe "basic", which only includes data from 2017, based on the dataset "df_2017"

In [12]:
basic = df_2017[df_2017['year'] == 2017]
basic.shape

(162, 120)

##### Looking for missing values

In [13]:
null = basic.columns[basic.isnull().any()]
nulist = basic[null].isnull().sum()
pd.set_option("display.max_rows", None)
nulist.sort_values(ascending=True)

ef_trade_tariffs                        1
ef_trade_movement_visit                 1
ef_trade_movement_capital               1
pf_movement_women                       1
pf_movement_foreign                     1
pf_religion                             1
pf_movement_domestic                    1
ef_legal_enforcement                    1
pf_ss_disappearances_disap              1
pf_identity_divorce                     1
pf_ss_women_inheritance                 2
ef_trade_regulatory_compliance          2
ef_regulation_credit_interest           2
ef_trade_tariffs_sd                     2
ef_legal_restrictions                   2
pf_religion_harassment                  2
pf_religion_restrictions                2
ef_trade_tariffs_mean                   2
ef_regulation_business_licensing        3
ef_trade_regulatory_nontariff           3
ef_regulation_labor_dismissal           3
ef_government_soa                       4
ef_government_tax_payroll               4
ef_legal_protection               

##### Deleting the 10 completly empty columns

In [14]:
basic.drop(["pf_association_sport_operate", "pf_association_sport_establish", "pf_association_prof_operate", "pf_association_prof_establish", "pf_association_political_operate", "pf_association_political_establish", "pf_religion_estop_operate", "pf_religion_estop_establish", "pf_ss_women_inheritance_daughters", "pf_ss_women_inheritance_widows"], axis = 1, inplace = True)

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:3990: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [15]:
basic.shape

(162, 110)

### What columns from the "freedom"-dataset do I put sure into my new dataset?

In [16]:
col_list_basic = list(basic)
print(col_list_basic)

['year', 'ISO_code', 'countries', 'region', 'hf_score', 'hf_rank', 'hf_quartile', 'pf_rol_procedural', 'pf_rol_civil', 'pf_rol_criminal', 'pf_rol', 'pf_ss_homicide', 'pf_ss_disappearances_disap', 'pf_ss_disappearances_violent', 'pf_ss_disappearances_organized', 'pf_ss_disappearances_fatalities', 'pf_ss_disappearances_injuries', 'pf_ss_disappearances', 'pf_ss_women_fgm', 'pf_ss_women_inheritance', 'pf_ss_women', 'pf_ss', 'pf_movement_domestic', 'pf_movement_foreign', 'pf_movement_women', 'pf_movement', 'pf_religion_estop', 'pf_religion_harassment', 'pf_religion_restrictions', 'pf_religion', 'pf_association_association', 'pf_association_assembly', 'pf_association_political', 'pf_association_prof', 'pf_association_sport', 'pf_association', 'pf_expression_killed', 'pf_expression_jailed', 'pf_expression_influence', 'pf_expression_control', 'pf_expression_cable', 'pf_expression_newspapers', 'pf_expression_internet', 'pf_expression', 'pf_identity_legal', 'pf_identity_sex_male', 'pf_identity_s

##### Most important features for my new dataset are:
- hf_score 
- hf_quartile
- countries
- region

##### Other features to use are:
- ISO_code
- hf_rank
- pf_score
- pf_rank
- ef_score
- ef_rank
- pf_religion
- pf_rol
- pf_ss_women
- pf_ss
- pf_expression
- pf_identity

### Dropping all other features to create the basic dataset

In [17]:
basic.drop(['year', 'pf_rol_procedural', 'pf_rol_civil', 'pf_rol_criminal', 'pf_ss_homicide', 
                 'pf_ss_disappearances_disap', 'pf_ss_disappearances_violent', 'pf_ss_disappearances_organized', 
                 'pf_ss_disappearances_fatalities', 'pf_ss_disappearances_injuries', 'pf_ss_disappearances', 
                 'pf_ss_women_fgm', 'pf_ss_women_inheritance', 'pf_movement_domestic', 'pf_movement_foreign', 
                 'pf_movement_women', 'pf_movement', 'pf_religion_estop', 'pf_religion_harassment', 
                 'pf_religion_restrictions', 'pf_association_association', 'pf_association_assembly', 
                 'pf_association_political', 'pf_association_prof', 'pf_association_sport', 'pf_association', 
                 'pf_expression_killed', 'pf_expression_jailed', 'pf_expression_influence', 'pf_expression_control', 
                 'pf_expression_cable', 'pf_expression_newspapers', 'pf_expression_internet', 'pf_identity_legal', 
                 'pf_identity_sex_male', 'pf_identity_sex_female', 'pf_identity_sex', 'pf_identity_divorce', 
                 'ef_government_consumption', 'ef_government_transfers', 'ef_government_enterprises', 
                 'ef_government_tax_income', 'ef_government_tax_payroll', 'ef_government_tax', 'ef_government_soa', 
                 'ef_government', 'ef_legal_judicial', 'ef_legal_courts', 'ef_legal_protection', 'ef_legal_military', 
                 'ef_legal_integrity', 'ef_legal_enforcement', 'ef_legal_restrictions', 'ef_legal_police', 'ef_legal_crime', 
                 'ef_legal_gender', 'ef_legal', 'ef_money_growth', 'ef_money_sd', 'ef_money_inflation', 'ef_money_currency', 
                 'ef_money', 'ef_trade_tariffs_revenue', 'ef_trade_tariffs_mean', 'ef_trade_tariffs_sd', 'ef_trade_tariffs', 
                 'ef_trade_regulatory_nontariff', 'ef_trade_regulatory_compliance', 'ef_trade_regulatory', 'ef_trade_black', 
                 'ef_trade_movement_foreign', 'ef_trade_movement_capital', 'ef_trade_movement_visit', 'ef_trade_movement', 
                 'ef_trade', 'ef_regulation_credit_ownership', 'ef_regulation_credit_private', 'ef_regulation_credit_interest', 
                 'ef_regulation_credit', 'ef_regulation_labor_minwage', 'ef_regulation_labor_firing', 
                 'ef_regulation_labor_bargain', 'ef_regulation_labor_hours', 'ef_regulation_labor_dismissal', 
                 'ef_regulation_labor_conscription', 'ef_regulation_labor', 'ef_regulation_business_adm', 
                 'ef_regulation_business_bureaucracy', 'ef_regulation_business_start', 'ef_regulation_business_bribes', 
                 'ef_regulation_business_licensing', 'ef_regulation_business_compliance', 'ef_regulation_business', 'ef_regulation'], axis = 1, inplace = True)

In [18]:
basic.head(3)

ISO_code countries                      region  hf_score  hf_rank  \
0      ALB   Albania              Eastern Europe      7.84     38.0   
1      DZA   Algeria  Middle East & North Africa      4.99    155.0   
2      AGO    Angola          Sub-Saharan Africa      5.40    151.0   

   hf_quartile  pf_rol  pf_ss_women  pf_ss  pf_religion  pf_expression  \
0          1.0     5.3          8.8    9.3          9.2            8.6   
1          4.0     3.8          5.0    7.8          4.9            7.3   
2          4.0     3.4          7.5    8.1          8.8            6.5   

   pf_identity  pf_score  pf_rank  ef_score  ef_rank  
0          5.8      8.01     46.0      7.67     30.0  
1          0.0      5.20    146.0      4.77    159.0  
2          5.0      5.98    121.0      4.83    158.0

In [19]:
basic.shape

(162, 16)

In [20]:
basic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 162 entries, 0 to 161
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ISO_code       162 non-null    object 
 1   countries      162 non-null    object 
 2   region         162 non-null    object 
 3   hf_score       162 non-null    float64
 4   hf_rank        162 non-null    float64
 5   hf_quartile    162 non-null    float64
 6   pf_rol         162 non-null    float64
 7   pf_ss_women    162 non-null    float64
 8   pf_ss          162 non-null    float64
 9   pf_religion    161 non-null    float64
 10  pf_expression  162 non-null    float64
 11  pf_identity    162 non-null    float64
 12  pf_score       162 non-null    float64
 13  pf_rank        162 non-null    float64
 14  ef_score       162 non-null    float64
 15  ef_rank        162 non-null    float64
dtypes: float64(13), object(3)
memory usage: 21.5+ KB


#### One missing value in the dataset (pf_religion for "Yemen")

In [21]:
basic[159:160]

ISO_code    countries                      region  hf_score  hf_rank  \
159      YEM  Yemen, Rep.  Middle East & North Africa       4.3    160.0   

     hf_quartile  pf_rol  pf_ss_women  pf_ss  pf_religion  pf_expression  \
159          4.0     2.4          4.1    4.2          NaN            5.0   

     pf_identity  pf_score  pf_rank  ef_score  ef_rank  
159          0.0      2.75    161.0      5.84    140.0

##### Because it is just one value (of hf_quartile 4.0), I fill it with the mean value of the countries in hf_quartile 4.0

In [22]:
fill_val = basic[basic['hf_quartile'] == 4.0]
fill_val.pf_religion.mean()

6.233333333333333

In [23]:
basic.fillna(6.23, inplace=True)

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:4147: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


In [24]:
basic[159:160]

ISO_code    countries                      region  hf_score  hf_rank  \
159      YEM  Yemen, Rep.  Middle East & North Africa       4.3    160.0   

     hf_quartile  pf_rol  pf_ss_women  pf_ss  pf_religion  pf_expression  \
159          4.0     2.4          4.1    4.2         6.23            5.0   

     pf_identity  pf_score  pf_rank  ef_score  ef_rank  
159          0.0      2.75    161.0      5.84    140.0

##### For better understanding I rename the Columns

In [25]:
basic.rename(columns={'countries': 'Country', 'region': 'Region', 'hf_score': 'FreeSc', 
                       'hf_rank': 'FreeRk', 'pf_rol': 'RoLaw', 'pf_ss_women': 'FemSafe', 
                       'pf_ss': 'PersSafe', 'pf_religion': 'RelFree', 'pf_expression': 'ExpFree', 
                       'pf_identity': 'GenderIdty', 'pf_score' : 'PersFreeSc', 'pf_rank': 'PersFreeRk', 
                       'ef_score': 'EcoFreeSc', 'ef_rank': 'EcoFreeRk', 'hf_quartile': 'FreeQuart'}, inplace=True)

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:4125: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


In [26]:
basic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 162 entries, 0 to 161
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ISO_code    162 non-null    object 
 1   Country     162 non-null    object 
 2   Region      162 non-null    object 
 3   FreeSc      162 non-null    float64
 4   FreeRk      162 non-null    float64
 5   FreeQuart   162 non-null    float64
 6   RoLaw       162 non-null    float64
 7   FemSafe     162 non-null    float64
 8   PersSafe    162 non-null    float64
 9   RelFree     162 non-null    float64
 10  ExpFree     162 non-null    float64
 11  GenderIdty  162 non-null    float64
 12  PersFreeSc  162 non-null    float64
 13  PersFreeRk  162 non-null    float64
 14  EcoFreeSc   162 non-null    float64
 15  EcoFreeRk   162 non-null    float64
dtypes: float64(13), object(3)
memory usage: 21.5+ KB


In [27]:
basic.head(3)

ISO_code  Country                      Region  FreeSc  FreeRk  FreeQuart  \
0      ALB  Albania              Eastern Europe    7.84    38.0        1.0   
1      DZA  Algeria  Middle East & North Africa    4.99   155.0        4.0   
2      AGO   Angola          Sub-Saharan Africa    5.40   151.0        4.0   

   RoLaw  FemSafe  PersSafe  RelFree  ExpFree  GenderIdty  PersFreeSc  \
0    5.3      8.8       9.3      9.2      8.6         5.8        8.01   
1    3.8      5.0       7.8      4.9      7.3         0.0        5.20   
2    3.4      7.5       8.1      8.8      6.5         5.0        5.98   

   PersFreeRk  EcoFreeSc  EcoFreeRk  
0        46.0       7.67       30.0  
1       146.0       4.77      159.0  
2       121.0       4.83      158.0

##### Rename some countries for better merging

In [28]:
basic.at[34,"Country"]= "Democratic Republic of the Congo"
basic.at[35,"Country"]= "Republic of the Congo"
basic.at[37,"Country"]= "Ivory Coast"
basic.at[42,"Country"]= "Dominican Republic"
basic.at[79,"Country"]= "South Korea"
basic.at[159,"Country"]= "Yemen"
basic.at[132,"Country"]= "Slovakia"
basic.at[40,"Country"]= "Czechia"
basic.at[81,"Country"]= "Kyrgyzstan"
basic.at[29,"Country"]= "Central African Republic"
basic.at[37,"Country"]= "Ivory Coast"
basic = basic.sort_values(by=['Country'], ascending=True)

### Now this dataset is ready and I will use it at the end as my basic dataframe.
### The next step is to prepare the other datasets for merging to add more features to complete my dataframe. 
##### Some countries have different notations in the datasets. I tried so solve this merging-issue with different solutions like pycountry_convert and the alpha2-code or mapping, but they all need the right notation too, so I must solve it manually for every dataset.
#### The first dataset is the Happiness-Index.
The scores and rankings utilize information from the Gallup World Survey. They are based on answers to the most life evaluation address, known as the Cantril step, asks respondents to think of a step with the most excellent conceivable life for them being a 10 and the most exceedingly bad conceivable life being a and to rate their claim current lives on that scale. This file contains the Happiness Score for 153 countries along with the factors used to explain the score.
The Happiness Score is explained by the following factors:
GDP per capita,
Healthy Life Expectancy,
Social support,
Freedom to make life choices,
Generosity,
Corruption Perception and the 
Residual error.

In [29]:
happy = pd.read_csv("files/World_HappinessReport2017.csv")
happy.head(3)
# Reserve Datasets: ("2017") is the same Data

Country  Happiness.Rank  Happiness.Score  Whisker.high  Whisker.low  \
0   Norway               1            7.537      7.594445     7.479556   
1  Denmark               2            7.522      7.581728     7.462272   
2  Iceland               3            7.504      7.622030     7.385970   

   Economy..GDP.per.Capita.    Family  Health..Life.Expectancy.   Freedom  \
0                  1.616463  1.533524                  0.796667  0.635423   
1                  1.482383  1.551122                  0.792566  0.626007   
2                  1.480633  1.610574                  0.833552  0.627163   

   Generosity  Trust..Government.Corruption.  Dystopia.Residual  
0    0.362012                       0.315964           2.277027  
1    0.355280                       0.400770           2.313707  
2    0.475540                       0.153527           2.322715

In [30]:
happy.shape

(155, 12)

In [31]:
col_list_happy = list(happy)
print(col_list_happy)

['Country', 'Happiness.Rank', 'Happiness.Score', 'Whisker.high', 'Whisker.low', 'Economy..GDP.per.Capita.', 'Family', 'Health..Life.Expectancy.', 'Freedom', 'Generosity', 'Trust..Government.Corruption.', 'Dystopia.Residual']


#### I will only use the (renamed) Happiness.Score and Happiness.Rank, because I will analyze them later with other features. Now dropping all other features.

In [32]:
happy.drop(['Whisker.high', 'Whisker.low', 'Economy..GDP.per.Capita.',
                'Family', 'Health..Life.Expectancy.', 'Freedom', 
                'Generosity', 'Trust..Government.Corruption.', 'Dystopia.Residual'], axis = 1, inplace = True)

In [33]:
happy.rename(columns={'Happiness.Rank': 'HappyRk', 'Happiness.Score': 'HappySc'}, inplace=True)

In [34]:
happy.shape

(155, 3)

In [35]:
happy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155 entries, 0 to 154
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Country  155 non-null    object 
 1   HappyRk  155 non-null    int64  
 2   HappySc  155 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 3.8+ KB


##### No missing values. Adjustment of Country-notation to the basic dataset.

In [36]:
happy.at[125,"Country"]= "Democratic Republic of the Congo"
happy.at[123,"Country"]= "Republic of the Congo"
happy.at[91,"Country"]= "North Macedonia"
happy.at[22,"Country"]= "Czechia"
happy.at[70,"Country"]= "Hong Kong"
happy.at[32,"Country"]= "Taiwan"
happy = happy.sort_values(by=['Country'], ascending=True)

#### After this Dataset is prepared for merging, going to the next dataset, which gives me the Country wise GDP.
GDP is the final value of the goods and services produced within the geographic boundaries of a country during a specified period of time, normally a year. GDP growth rate is an important indicator of the economic performance of a country. The data is from World Bank and OECD.

In [37]:
gdp1 = pd.read_csv("files//Country wise GDP from 1994 to 2017.csv")
gdp1 = gdp1[gdp1['Year'] == 2017]
gdp = gdp1.sort_values('Country', ascending=True)
gdp.head(3)

Year      Country  GDP (in USD)  GDP Real (in USD)  GDP change (%)  \
2608  2017  Afghanistan   19543976895        20744935406            2.67   
2866  2017      Albania   13038538300        13986932579            3.84   
1238  2017      Algeria  167555280113       199171379146            1.60   

      GDP per capita (in USD)  Pop. change (%)  Population  
2608                      572             2.58    36296113  
2866                     4850            -0.08     2884169  
1238                     4812             2.07    41389189

In [38]:
gdp.shape

(189, 8)

In [39]:
col_list_gdp = list(gdp)
print(col_list_gdp)

['Year', 'Country', 'GDP (in USD)', 'GDP Real (in USD)', 'GDP change (%)', 'GDP per capita (in USD)', 'Pop. change (%)', 'Population']


#### You can't compare the "GDP" itself because of different Population-Count, so I will use the "GDP per capita".
The gross domestic product per capita, or GDP per capita, is a measure of a country's economic output that accounts for its number of people. It divides the country's gross domestic product by its total population.

#### After comparing "Population" with other Population-Count-datasets, I saw I can take it from here.
("FAOSTAT_data_6-13-2019.csv", "countries of the world.csv", "population_total_long")
##### Dropping all other features and rename one feature.

In [40]:
gdp.drop(['Year', 'GDP (in USD)', 'GDP Real (in USD)', 'GDP change (%)', 'Pop. change (%)'], axis = 1, inplace = True)
gdp.rename(columns={'GDP per capita (in USD)': 'GDP_per_capita'}, inplace=True)
gdp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 189 entries, 2608 to 2536
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Country         189 non-null    object
 1   GDP_per_capita  189 non-null    int64 
 2   Population      189 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 5.9+ KB


##### No missing values. Adjustment of Country-notation to the basic dataset.

In [41]:
gdp.at[2148,"Country"]= "Democratic Republic of the Congo"
gdp.at[3226,"Country"]= "Republic of the Congo"
gdp.at[1598,"Country"]= "Dominican Republic"
gdp.at[1104,"Country"]= "Czechia"
gdp.at[2196,"Country"]= "Ivory Coast"
gdp = gdp.sort_values('Country', ascending=True)

#### Taking the Life-expectancy (in years) from the next dataset
##### Original-Data is from the WHO

In [42]:
life_ex = pd.read_csv("files/life-expectancy.csv")
life_ex = life_ex[life_ex['Year'] == 2017]
life_ex.rename(columns={"Entity": "Country", "Life expectancy (years)" : "LifeAge"}, inplace=True)
life_ex.drop(['Year', 'Code'], axis = 1, inplace = True)
life_ex.head(3)

Country  LifeAge
67   Afghanistan   64.130
139       Africa   62.472
209      Albania   78.333

In [43]:
life_ex.shape

(243, 2)

In [44]:
life_ex.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 243 entries, 67 to 19025
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Country  243 non-null    object 
 1   LifeAge  243 non-null    float64
dtypes: float64(1), object(1)
memory usage: 5.7+ KB


##### No missing values. Adjustment of Country-notation to the basic dataset.

In [45]:
life_ex.at[4428,"Country"]= "Democratic Republic of the Congo"
life_ex.at[3770,"Country"]= "Republic of the Congo"
life_ex.at[4827,"Country"]= "Dominican Republic"
life_ex.at[4358,"Country"]= "Czechia"
life_ex.at[9901,"Country"]= "North Macedonia"
life_ex.at[3986,"Country"]= "Ivory Coast"
life_ex = life_ex.sort_values(by=['Country'], ascending=True)

#### Dataset "rel_imp" for measure of religiousity-importance per country (in %)

##### Religiosity = percentage of people claiming themselves as religious

In [46]:
rel_imp = pd.read_csv("files/religion_vs_GDP_per_Capita.csv")
rel_imp.rename(columns={"country": "Country", "religiousity%" : "RelImp"}, inplace=True)
rel_imp.drop(['Unnamed: 0', 'US$'], axis = 1, inplace = True)
rel_imp.head(3)

Country  RelImp
0    Angola      88
1    Brazil      79
2  Bulgaria      52

In [47]:
rel_imp.shape

(148, 2)

In [48]:
rel_imp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148 entries, 0 to 147
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Country  148 non-null    object
 1   RelImp   148 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 2.4+ KB


##### No missing values. Adjustment of Country-notation to the basic dataset.

In [49]:
rel_imp.at[73,"Country"]= "Czechia"
rel_imp.at[141,"Country"]= "North Macedonia"
rel_imp = rel_imp.sort_values(by=['Country'], ascending=True)

### Dataset "peace" for World Peace Index. 
##### Data from the Institute for Economics and Peace. The values has to be read with scepticismn, so I just use the ranking. 

In [50]:
peace1 = pd.read_csv("files/gpi-2008-2019.csv")
peace = peace1.sort_values('Country', ascending=True)
peace = peace[["Country", "2017 rank"]]
peace.rename(columns={"2017 rank": "PeaceRk"}, inplace=True)
peace.head(3)

Country  PeaceRk
157  Afghanistan      159
50       Albania       60
110      Algeria      107

In [51]:
peace.shape

(163, 2)

In [52]:
peace.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 163 entries, 157 to 131
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Country  163 non-null    object
 1   PeaceRk  163 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 3.8+ KB


##### No missing values. Adjustment of Country-notation to the basic dataset.

In [53]:
peace.at[94,"Country"]= "Kyrgyzstan"
peace.at[106,"Country"]= "Ivory Coast"
peace = peace.sort_values(by=['Country'], ascending=True)

#### Dataset "corrupt" for World Corruption Index.
From Transparency International: The CPI scores and ranks countries/territories based on how corrupt a country’s public sector is perceived to be. It is a composite index, a combination of surveys and assessments of corruption, collected by a variety of reputable institutions. The CPI is the most widely used indicator of corruption worldwide. Higher score means less corruption.

In [54]:
corrupt1 = pd.read_csv("files/CPI_2017_final_dataset.csv")
corrupt = corrupt1.sort_values('Country', ascending=True)
corrupt = corrupt[["Country", "CPI Score 2017", "Rank 2017"]]
corrupt.rename(columns={"CPI Score 2017": "CorruptSc", "Rank 2017": "CorruptRk"}, inplace=True)
corrupt.head(3)

Country  CorruptSc  CorruptRk
176  Afghanistan         15        177
93       Albania         38         91
112      Algeria         33        112

In [55]:
corrupt.shape

(180, 3)

In [56]:
corrupt.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 180 entries, 176 to 156
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Country    180 non-null    object
 1   CorruptSc  180 non-null    int64 
 2   CorruptRk  180 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 5.6+ KB


##### No missing values. Adjustment of Country-notation to the basic dataset.

In [57]:
corrupt.at[50,"Country"]= "South Korea"
corrupt.at[43,"Country"]= "Czechia"
corrupt.at[15,"Country"]= "United States"
corrupt.at[160,"Country"]= "Republic of the Congo"
corrupt.at[102,"Country"]= "Ivory Coast"
corrupt.at[172,"Country"]= "North Korea"
corrupt.at[108,"Country"]= "North Macedonia"
corrupt = corrupt.sort_values(by=['Country'], ascending=True)

#### Dataset Gender Inequality Index (GII)

Original-Data is from the United Nations Development Programme. Gender inequality remains a major barrier to human development. Girls and women have made major strides since 1990, but they have not yet gained gender equity. The disadvantages facing women and girls are a major source of inequality. All too often, women and girls are discriminated against in health, education, political representation, labour market, etc.—with negative consequences for development of their capabilities and their freedom of choice.

The GII is an inequality index. It measures gender inequalities in three important aspects of human development—reproductive health, measured by maternal mortality ratio and adolescent birth rates; empowerment, measured by proportion of parliamentary seats occupied by females and proportion of adult females and males aged 25 years and older with at least some secondary education; and economic status, expressed as labour market participation and measured by labour force participation rate of female and male populations aged 15 years and older. The GII is built on the same framework as the IHDI—to better expose differences in the distribution of achievements between women and men. It measures the human development costs of gender inequality. Thus the higher the GII value the more disparities between females and males and the more loss to human development.

The GII sheds new light on the position of women in 162 countries; it yields insights in gender gaps in major areas of human development. The component indicators highlight areas in need of critical policy intervention and it stimulates proactive thinking and public policy to overcome

In [58]:
gie = pd.read_csv("files/Gender Inequality Index (1).csv")
gie = gie[["Country", "2017"]]
gie.rename(columns={"2017": "GendEqual"}, inplace=True)
gie.head(3)

Country  GendEqual
0   Afghanistan      0.653
1       Albania      0.238
2       Algeria      0.442

In [59]:
gie.shape

(161, 2)

In [60]:
gie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 161 entries, 0 to 160
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Country    160 non-null    object 
 1   GendEqual  157 non-null    float64
dtypes: float64(1), object(1)
memory usage: 2.6+ KB


##### I took the missing values for Chile (0.720), Côte d'Ivoire (0.841) and Honduras(0.479) from the Internet and drop the last row, which has no values.

In [61]:
gie.at[30,"GendEqual"]= 0.720 # Chile
gie.at[40,"GendEqual"]= 0.841 # Côte d'Ivoire
gie.at[61,"GendEqual"]= 0.479 # Honduras
gie = gie[:-1] 

In [62]:
gie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 160 entries, 0 to 159
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Country    160 non-null    object 
 1   GendEqual  160 non-null    float64
dtypes: float64(1), object(1)
memory usage: 2.6+ KB


##### Deleting the first sign (spaces) of the countries

In [63]:
cols_to_check = ['Country']
for col in cols_to_check:
    gie[col] = gie[col].apply(lambda x : x[1:] if x.startswith(" ") else x)

##### No missing values. Adjustment of Country-notation to the basic dataset.

In [64]:
gie.at[33,"Country"]= "Republic of the Congo"
gie.at[34,"Country"]= "Democratic Republic of the Congo"
gie.at[76,"Country"]= "South Korea"
gie.at[142,"Country"]= "North Macedonia"
gie.at[156,"Country"]= "Vietnam"
gie.at[66,"Country"]= "Iran"
gie.at[40,"Country"]= "Ivory Coast"
gie.at[95,"Country"]= "Moldova"
gie.at[119,"Country"]= "Russia"
gie.at[155,"Country"]= "Venezuela"
gie = gie.sort_values(by=['Country'], ascending=True)

### The Human Development Index (HDI) 
A statistic composite index - also from the United Nations Development Programme - of life expectancy, education (Literacy Rate, Gross Enrollment Ratio at different levels and Net Attendance Ratio), and per capita income indicators, which are used to rank countries into four tiers of human development. A country scores a higher HDI when the lifespan is higher, the education level is higher, and the gross national income GNI (PPP) per capita is higher.

In [65]:
hdi = pd.read_csv("files/Human Development Index (HDI).csv")
hdi = hdi.sort_values('Country', ascending=True)
hdi = hdi[["Country", "HDI Rank", "2017"]]
hdi.rename(columns={"HDI Rank": "HumDevRk", "2017": "HumDevSc"}, inplace=True)
hdi.tail(25)

Country  \
184                                            Vanuatu   
185                 Venezuela (Bolivarian Republic of)   
186                                           Viet Nam   
187                                              Yemen   
188                                             Zambia   
189                                           Zimbabwe   
197                                        Arab States   
195                               Developing Countries   
198                          East Asia and the Pacific   
199                            Europe and Central Asia   
192                             High human development   
190                                  Human Development   
200                    Latin America and the Caribbean   
203                          Least Developed Countries   
194                              Low human development   
193                           Medium human development   
205  Organization for Economic Co-operation and Dev...   
196                                            Regions   
204                     Small Island Developing States   
201                                         South Asia   
202                                 Sub-Saharan Africa   
191                        Very high human development   
206                                              World   
0                                                  NaN   
207                                                NaN   

                                              HumDevRk HumDevSc  
184                                                141    0.595  
185                                                 96    0.735  
186                                                118    0.690  
187                                                177    0.463  
188                                                143    0.589  
189                                                150    0.553  
197                                                       0.701  
195                                                       0.683  
198                                                       0.737  
199                                                       0.776  
192                                                       0.746  
190                                                NaN       ..  
200                                                       0.758  
203                                                       0.525  
194                                                       0.505  
193                                                       0.630  
205                                                       0.894  
196                                                NaN       ..  
204                                                       0.722  
201                                                       0.639  
202                                                       0.539  
191                                                       0.890  
206                                                       0.729  
0    Human Development Index (HDI)\nnull\nDimension...      NaN  
207                                                NaN      NaN

##### Dropping the (for me) useless Data from Index 189 on 

In [66]:
hdi = hdi[0:189] 
hdi.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 189 entries, 1 to 189
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Country   189 non-null    object
 1   HumDevRk  189 non-null    object
 2   HumDevSc  189 non-null    object
dtypes: object(3)
memory usage: 5.9+ KB


##### Change Datatype of numeric Columns from Object

In [67]:
b = list(hdi.columns.values) 
del b[0] # "Country"
def changetyp(i): # Function to change Datatype for both numeric columns
    for i in b:
        hdi[i] = pd.to_numeric(hdi[i], errors='coerce')
changetyp(b)
hdi.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 189 entries, 1 to 189
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Country   189 non-null    object 
 1   HumDevRk  189 non-null    int64  
 2   HumDevSc  189 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 5.9+ KB


##### Deleting the first sign (spaces) of the countries

In [68]:
cols_to_check = ['Country']
for col in cols_to_check:
    hdi[col] = hdi[col].apply(lambda x : x[1:] if x.startswith(" ") else x)

In [69]:
c_names_diff = []
for i in hdi['Country'].unique():
    if i not in basic['Country'].unique():
        c_names_diff.append(i)
print(c_names_diff)

['Afghanistan', 'Andorra', 'Antigua and Barbuda', 'Bolivia (Plurinational State of)', 'Cabo Verde', 'Comoros', 'Congo', 'Congo (Democratic Republic of the)', 'Cuba', "Côte d'Ivoire", 'Djibouti', 'Dominica', 'Equatorial Guinea', 'Eritrea', 'Eswatini (Kingdom of)', 'Gambia', 'Grenada', 'Hong Kong, China (SAR)', 'Iran (Islamic Republic of)', 'Kiribati', 'Korea (Republic of)', "Lao People's Democratic Republic", 'Liechtenstein', 'Maldives', 'Marshall Islands', 'Micronesia (Federated States of)', 'Moldova (Republic of)', 'Palau', 'Palestine, State of', 'Papua New Guinea', 'Russian Federation', 'Saint Kitts and Nevis', 'Saint Lucia', 'Saint Vincent and the Grenadines', 'Samoa', 'Sao Tome and Principe', 'Solomon Islands', 'South Sudan', 'Syrian Arab Republic', 'Tanzania (United Republic of)', 'Tonga', 'Turkmenistan', 'Uzbekistan', 'Vanuatu', 'Venezuela (Bolivarian Republic of)', 'Viet Nam']


In [70]:
hdi.at[40,"Country"]= "Democratic Republic of the Congo"
hdi.at[39,"Country"]= "Republic of the Congo"
hdi.at[114,"Country"]= "Moldova"
hdi.at[91,"Country"]= "South Korea"
hdi.at[142,"Country"]= "Russia"
hdi.at[186,"Country"]= "Vietnam"
hdi.at[46,"Country"]= "Ivory Coast"
hdi.at[168,"Country"]= "Tanzania"
hdi.at[185,"Country"]= "Venezuela"
hdi.at[75,"Country"]= "Hong Kong"
hdi.at[166,"Country"]= "Syria"
hdi.at[21,"Country"]= "Bolivia"
hdi.at[80,"Country"]= "Iran"
hdi = hdi.sort_values(by=['Country'], ascending=True)

### The Datasets are prepared for merging

##### First I wanted to put six countries, missing by me in the basic dataset and rename it in "df. After merging the other datasets this will be my dataframe. But I decided against these countries because they had to much missing values.

In [71]:
#mergehelp = life_ex.T
#mergehelp2 = mergehelp[[67, 4137, 12309, 17373, 18313]] # 'Afghanistan', 'Cuba', 'North Korea', 'Turkmenistan', 'Uzbekistan'
#m3 = mergehelp2.T
#m3 = m3[["Country"]]
#df1 = pd.merge(basic, m3, on='Country', how = "outer")
#df1 = df1.sort_values(by=['Country'], ascending=True)
#print("basic-shape = " + str(basic.shape))
#print("Filling-shape = " + str(m3.shape))
#print("df-shape = " + str(df1.shape))
#df1.info()

### Now merging the 8 other Datasets to my dataframe 

In [72]:
df = pd.merge(basic, happy, on="Country", how = "left")
df = pd.merge(df, gdp, on='Country', how = "left")
df = pd.merge(df, life_ex, on='Country', how = "left")
df = pd.merge(df, rel_imp, on='Country', how = "left")
df = pd.merge(df, peace, on='Country', how = "left")
df = pd.merge(df, corrupt, on='Country', how = "left")
df = pd.merge(df, gie, on='Country', how = "left")
df = pd.merge(df, hdi, on='Country', how = "left")
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 162 entries, 0 to 161
Data columns (total 28 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ISO_code        162 non-null    object 
 1   Country         162 non-null    object 
 2   Region          162 non-null    object 
 3   FreeSc          162 non-null    float64
 4   FreeRk          162 non-null    float64
 5   FreeQuart       162 non-null    float64
 6   RoLaw           162 non-null    float64
 7   FemSafe         162 non-null    float64
 8   PersSafe        162 non-null    float64
 9   RelFree         162 non-null    float64
 10  ExpFree         162 non-null    float64
 11  GenderIdty      162 non-null    float64
 12  PersFreeSc      162 non-null    float64
 13  PersFreeRk      162 non-null    float64
 14  EcoFreeSc       162 non-null    float64
 15  EcoFreeRk       162 non-null    float64
 16  HappyRk         147 non-null    float64
 17  HappySc         147 non-null    flo

In [73]:
#df.at[0,"ISO_code"]= "AFG"
#df.at[37,"ISO_code"]= "CUB"
#df.at[111,"ISO_code"]= "PRK"
#df.at[154,"ISO_code"]= "TKM"
#df.at[161,"ISO_code"]= "UZB"
#df.at[0,"Region"]= "South Asia"
#df.at[37,"Region"]= "Latin America & the Caribbean"
#df.at[111,"Region"]= "East Asia"
#df.at[154,"Region"]= "Caucasus & Central Asia"
#df.at[161,"Region"]= "Caucasus & Central Asia"

In [74]:
df.shape

(162, 28)

#### In every score of the dataframe the higher value is the "better" value except in Gender_equality, so I change it by creating a new column

In [75]:
df.eval('Gend_Eq = (1 - GendEqual) * 10', inplace = True)
df.drop('GendEqual', inplace=True, axis=1)

In [76]:
df.head(3)

ISO_code  Country                      Region  FreeSc  FreeRk  FreeQuart  \
0      ALB  Albania              Eastern Europe    7.84    38.0        1.0   
1      DZA  Algeria  Middle East & North Africa    4.99   155.0        4.0   
2      AGO   Angola          Sub-Saharan Africa    5.40   151.0        4.0   

   RoLaw  FemSafe  PersSafe  RelFree  ExpFree  GenderIdty  PersFreeSc  \
0    5.3      8.8       9.3      9.2      8.6         5.8        8.01   
1    3.8      5.0       7.8      4.9      7.3         0.0        5.20   
2    3.4      7.5       8.1      8.8      6.5         5.0        5.98   

   PersFreeRk  EcoFreeSc  EcoFreeRk  HappyRk  HappySc  GDP_per_capita  \
0        46.0       7.67       30.0    109.0    4.644          4850.0   
1       146.0       4.77      159.0     53.0    5.872          4812.0   
2       121.0       4.83      158.0    140.0    3.795          3410.0   

   Population  LifeAge  RelImp  PeaceRk  CorruptSc  CorruptRk  HumDevRk  \
0   2884169.0   78.333    39.0     60.0       38.0       91.0      69.0   
1  41389189.0   76.499    90.0    107.0       33.0      112.0      82.0   
2  29816766.0   60.379    88.0     74.0       19.0      167.0     149.0   

   HumDevSc  Gend_Eq  
0     0.789     7.62  
1     0.758     5.58  
2     0.576      NaN

In [77]:
df.to_csv("df.csv")

### The dataframe "df" now is ready for the EDA - see EDA-notebook